priam/src/main/java/com/netflix/priam/backup/AbstractBackup.java

```diff
@@ -80,48 +74,62 @@ private AbstractBackupPath getAbstractBackupPath(final File file, final BackupFi
-
-        File[] files = parent.listFiles();
-        if (files == null) return bps;
-
-        for (File file : files) {
-            if (file.isFile() && file.exists()) {
-                AbstractBackupPath bp = getAbstractBackupPath(file, type);
-
-                if (async)
-                    futures.add(
-                            fs.asyncUploadFile(
-                                    Paths.get(bp.getBackupFile().getAbsolutePath()),
-                                    Paths.get(bp.getRemotePath()),
-                                    bp,
-                                    10,
-                                    true));
-                else
-                    fs.uploadFile(
-                            Paths.get(bp.getBackupFile().getAbsolutePath()),
-                            Paths.get(bp.getRemotePath()),
-                            bp,
-                            10,
-                            true);
-
-                bps.add(bp);
-            }
+        ImmutableSet<AbstractBackupPath> bps = getBackupPaths(parent, type);
+        final List<Future<AbstractBackupPath>> futures = Lists.newArrayList();
+        for (AbstractBackupPath bp : bps) {
+            if (async) futures.add(fs.asyncUploadAndDelete(bp, 10));
+            else fs.uploadAndDelete(bp, 10);
         }
 
         // Wait for all files to be uploaded.
         if (async && waitForCompletion) {
-            for (Future future : futures)
+            for (Future<AbstractBackupPath> future : futures)
                 future.get(); // This might throw exception if there is any error
         }
 
         return bps;
     }
 
+    protected ImmutableSet<AbstractBackupPath> getBackupPaths(File dir, BackupFileType type)
+            throws IOException {
+        Set<File> files =
+                Files.list(dir.toPath()).map(Path::toFile).filter(File::isFile).collect(toSet());
+        Set<String> compressedFilePrefixes =
+                files.stream()
+                        .map(File::getName)
+                        .filter(name -> name.endsWith(COMPRESSION_SUFFIX))
+                        .map(name -> name.substring(0, name.lastIndexOf('-')))
+                        .collect(toSet());
+        final ImmutableSet.Builder<AbstractBackupPath> bps = ImmutableSet.builder();
+        for (File file : files) {
+            final AbstractBackupPath bp = pathFactory.get();
+            bp.parseLocal(file, type);
+            bp.setCompression(getCompressionAlgorithm(file.getName(), compressedFilePrefixes));
+            bps.add(bp);
+        }
+        return bps.build();
+    }
+
+    private CompressionAlgorithm getCompressionAlgorithm(String file, Set<String> compressedFiles) {
+        BackupsToCompress which = config.getBackupsToCompress();
+        switch (which) {
+            case NONE:
+                return CompressionAlgorithm.NONE;
+            case ALL:
+                return CompressionAlgorithm.SNAPPY;
+            case UNCOMPRESSED:
```

Comment:
nit: Can we name it more like IF_REQUIRED. Uncompressed does not show true intent.

Reply:
Also thinking aloud, they are not compressionAlgorithms (like snappy/lz4) but compressionType.

Reply:
I oppose this actually. IF_REQUIRED does not say which backups should be compressed. Is it a function of the file's size? NIC throughput? Number of cores?

I initially had it as:

Priam.whenToCompressBackups
enum WhenToCompress {
  ALWAYS
  NEVER
  WHEN_UNCOMPRESSED
}

but it led to fatter code for what I considered to be a minimal readability improvement. Let me know if you prefer that or insist upon IF_REQUIRED.

Reply:
In your second comment, Are you saying the CompressionAlgorithm enum should be renamed?

Reply:
I prefer elaborate rather than cryptic "uncompressed" which could mean to the operator that it is uncompressed rather than choosing NONE. And yes, I prefer renaming the enum as compression algo generally refer to SNAPPY, LZ4, xxHash etc.